### DISCUSSION TITLE
Mechanism to early-exit reactivity

### DISCUSSION THREAD

**Discussion by fonsp** [fonsp/Pluto.jl#1743] [Pluto development]
We now have `@use_memo` from https://github.com/JuliaPluto/PlutoHooks.jl , which allows me to memoize computations:

https://user-images.githubusercontent.com/6933510/145025981-817201aa-b336-4840-b32a-18fbab84a212.mov

This is great! But!

Even though the calculation of `result` is being memoized, it is still being redefined every time `x` changes, triggering a reactive run. If I use `result` throughout my notebook (doing further analysis), I now need to put `@use_memo` around everything that depends on `result`. This came up in my Catalyst.jl project and in #1706.

To solve this, we need a mechanism for notebook code to tell Pluto to **early-exit** reactivity. If we do this based on `===` equality to the last value, then it resembles the runtime of systems like https://www.youtube.com/watch?v=5GGl3TojN2o and https://github.com/nextjournal/clerk . We can't simply enable this by default, because there are really cool notebooks that depend on symbolic-based triggers (like https://www.youtube.com/watch?v=6Qs5EXDpZBI).

A slightly related issue is #1696 , though we can also just implement that without this generality

**Comment by fonsp** [comment]
# `@purevalue` macro

I imagine that we could do something like:
```julia
@bind x Slider(1:100)

z = x ÷ 20

result = @purevalue(@use_memo([z]) do
	expensive_calculation(z)
end)
```

Which tells Pluto that `result` is "pure": if its value did not change, it should not trigger reactivity.

Both `@purevalue` and `@use_memo` are low-level, but I imagine that we can combine the two into a single macro in PlutoLinks.jl:

```julia
result = @pure expensive_calculation(z)
```
where we automatically find dependencies using ExpressionExplorer (we copy-paste it into PlutoLinks.jl), or you can give the dependencies manually to allow tricks:
```julia
result = @pure expensive_calculation(z + x) [z]
```

(Though `@pure` is already taken by `Base.@pure`...)

**Reply by Pangoraw** [reply]
What should happen when a cell contain both pure and unpure variables ?

```julia
begin
  y = rand(x)
  z = @pure expensive_calculation(x)
end
```

Maybe the reactive run can stop all runs to the `:z` edges but not on `:y` ?

**Reply by fonsp** [reply]
Yes! It should apply to the edges!
